Update dependant nodes for ODP

sorja opened the pull request:
# Todo:
- [x] Merge 1a and 1b node updates to avoid double update for node

Resolves #1627

https://user-images.githubusercontent.com/5508251/198965817-238c2d22-20b0-48fc-a3f1-785ddaa3ba77.mov

Resolves #1666

https://user-images.githubusercontent.com/5508251/199221045-90919871-dc32-4a8d-9a87-0f669f74c685.mov

sorja committed:
initial commit for update odp

sorja committed:
Merge branch 'master-meta' of github.com:openforis/fra-platform into meta/update-dependants-on-odp-update

sorja committed:
use tablenames

sorja committed:
Forest Characteristics: Fix issue with column names

sorja committed:
Merge branch 'master-meta' into meta/update-dependants-on-odp-update

sorja committed:
Merge branch 'meta/fix-forest-characteristics-odp' into meta/update-dependants-on-odp-update

sorja committed:
Merge branch 'master-meta' of github.com:openforis/fra-platform into meta/update-dependants-on-odp-update

sorja committed:
integration test: Fix odp edit test

mergify[bot] committed:
Merge branch 'master-meta' into meta/update-dependants-on-odp-update

sorja committed:
Update 1a, 1b on odp change (#1667)

* Update 1a, 1b on odp change

* integration test: Mock socketService

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

sorja commented on src/client/store/pages/assessmentSection/slice.ts:
```diff
@@ -37,8 +37,19 @@ export const assessmentSectionSlice = createSlice({
-      const dataUpdate = TableDatas.updateDatum({ data, countryIso, tableName, variableName, colName, value })
-      state.data = { ...data, ...dataUpdate }
+      const dataUpdateProps = { data, countryIso, tableName, variableName, colName, value }
+      if (value.odp) {
+        // If the value overrides ODP value, update the ODP state
+        const dataUpdate = TableDatas.updateDatum({
+          ...dataUpdateProps,
+          data: state.originalDataPointData,
+          tableName: 'originalDataPointValue',
+        })
+        state.originalDataPointData = { ...state.originalDataPointData, ...dataUpdate }
+      } else {
+        const dataUpdate = TableDatas.updateDatum(dataUpdateProps)
+        state.data = { ...data, ...dataUpdate }
+      }
```
When we update->receive an ODP value from websocket, handle it here

Although, when writing this and working with the websockets, I was wondering should we just use one state variable (eg. `data` here) and receive the data directly from backend?

Do we ever use non-odp values for odp years when 1b has ODPs enabled?

minotogna replied:
this should not be changed. backend should pass something like:
["FIN-fra-2020-originalDataPointValue-otherWoodedLand-2020-nodeUpdate",[{"value":{"raw":"65346345.00","estimated":true,"validation":{"valid":false,"messages":[{"key":"extentOfForest.fedAreasExceedTotalLandArea"}]}}}]]

sorja replied:
yes

sorja commented on src/server/controller/cycleData/persistNodeValue/calculateDependantNodes/calculateDependantNodes.ts:
```diff
@@ -7,7 +7,10 @@ import { RowRepository } from '@server/repository/assessment/row'
 
-export const calculateDependantNodes = async (props: Props, client: BaseProtocol): Promise<NodeUpdates> => {
+export const calculateDependantNodes = async (
+  props: Omit<Props, 'value'>,
```
Omit property `value` as it was not used and we did not have it in updateOriginalDataPoint

minotogna commented on src/server/controller/cycleData/updateOriginalDataPoint.ts:
```diff
@@ -16,11 +185,34 @@ export const updateOriginalDataPoint = async (
+      user,
+      updatedOriginalDataPoint,
+      client: t,
+    })
```
there should be only one method invoked

sorja replied:
yes

minotogna commented on src/server/controller/cycleData/updateOriginalDataPoint.ts:
```diff
@@ -1,9 +1,178 @@
+  updatedOriginalDataPoint: OriginalDataPoint
+  client: BaseProtocol
+}
+
```
this is unreadable. extract methods

sorja replied:
yes

sorja committed:
Remove assessmentSection.originalDataPointData

sorja committed:
Update 1a, 1b on change

Merge node updates
Use single state entry (state.data)
ODP cell listen on odp changes

sorja committed:
Merge branch 'master-meta' into meta/update-dependants-on-odp-update

sorja committed:
deepscan

sorja committed:
Merge branch 'meta/update-dependants-on-odp-update' of github.com:openforis/fra-platform into meta/update-dependants-on-odp-update

minotogna commented on src/client/pages/AssessmentSection/DataTable/Table/Row/RowData/Cell/Cell.tsx:
```diff
@@ -66,7 +66,16 @@ const Cell: React.FC<Props> = (props) => {
+    countryIso,
+    assessmentName,
+    cycleName,
+    tableName: nodeValue?.odp ? 'originalDataPointValue' : tableName,
```
don't we have a contact for 'originalDataPointValue'?

minotogna commented on src/client/store/pages/assessmentSection/hooks/index.ts:
```diff
@@ -15,9 +15,7 @@ export const useTableSections = (props: { sectionName: string }): Array<TableSec
-  return useAppSelector(
-    (state) => state.pages.assessmentSection.originalDataPointData?.[countryIso]?.originalDataPointValue
-  )
+  return useAppSelector((state) => state.pages.assessmentSection.data?.[countryIso]?.originalDataPointValue)
```
I would use a constant here as well

minotogna commented on src/server/controller/cycleData/updateOriginalDataPoint/getForestCharacteristicsDependantUpdates.ts:
```diff
@@ -0,0 +1,98 @@
+    ),
+  }
+
+  const tableName = 'originalDataPointValue'
```
use a constant

minotogna commented on src/server/controller/cycleData/updateOriginalDataPoint/index.ts:
```diff
@@ -0,0 +1,71 @@
+    const colName = String(updatedOriginalDataPoint.year)
+    const { countryIso } = updatedOriginalDataPoint
+
+    const extentOfForestUpdates = await getExtentOfForestDependantUpdates({
+      countryIso,
+      assessment,
+      cycle,
+      colName,
+      user,
+      updatedOriginalDataPoint,
+      client: t,
+    })
+
+    const forestCharacteristicsUpdates = await getForestCharacteristicsDependantUpdates({
+      countryIso,
+      assessment,
+      cycle,
+      colName,
+      user,
+      updatedOriginalDataPoint,
+      client: t,
+    })
+
```
can't we call only one method ?

minotogna commented on src/server/controller/cycleData/updateOriginalDataPoint/updateDependantsProps.ts:
```diff
@@ -0,0 +1,15 @@
+  cycle: Cycle
+  user: User
+  colName: string
+  updatedOriginalDataPoint: OriginalDataPoint
```
can't this be called originalDataPoint directly ? also, use upper case for types

sorja committed:
Code quality

mergify[bot] committed:
Merge branch 'master-meta' into meta/update-dependants-on-odp-update

mergify[bot] committed:
Merge branch 'master-meta' into meta/update-dependants-on-odp-update

minotogna commented on src/server/controller/cycleData/updateOriginalDataPoint/index.ts:
```diff
@@ -16,11 +17,13 @@ export const updateOriginalDataPoint = async (
-      { assessment, cycle, originalDataPoint },
-      t
-    )
+    await updateDependantNodes({ originalDataPoint: updatedOriginalDataPoint, assessment, cycle, user }, t)
```
cant we just reuse the logic from src/server/controller/cycleData/persistNodeValue/index.ts line 18-31? why reimplementing the logic here?

minotogna replied:
we should talk in person

minotogna commented on src/test/integration/assessment/originalDataPoint.ts:
```diff
@@ -47,7 +50,11 @@ export default (): void =>
-        assessment,
+        assessment: {
+          ...assessment,
+          // Test assessment meta cache is empty
+          metaCache: { ...assessment.metaCache, calculations: { dependants: {}, dependencies: {} } },
+        },
```
cant this be added to the test assessment directly ?

sorja replied:
assessment fetched from db, should we put it in db..?

https://github.com/openforis/fra-platform/blob/e2031fc11b7e832653914f6545f59a1081fce7ff/src/test/integration/assessment/originalDataPoint.ts#L21

minotogna commented on src/server/controller/cycleData/persistNodeValue/calculateDependantNodes/calculateDependantNodes.ts:
```diff
@@ -7,7 +7,10 @@ import { RowRepository } from '@server/repository/assessment/row'
 
-export const calculateDependantNodes = async (props: Props, client: BaseProtocol): Promise<NodeUpdates> => {
+export const calculateDependantNodes = async (
+  props: Omit<Props, 'value'>,
```
check if value is not needed in src/server/controller/cycleData/persistNodeValue/index.ts

sorja replied:
It is not used
https://github.com/openforis/fra-platform/blob/e2031fc11b7e832653914f6545f59a1081fce7ff/src/server/controller/cycleData/persistNodeValue/index.ts#L18

minotogna commented on src/meta/assessment/table.ts:
```diff
@@ -14,6 +14,7 @@ export enum TableNames {
   primaryDesignatedManagementObjective = 'primaryDesignatedManagementObjective',
 
   valueAggregate = 'value_aggregate',
+  originalDataPointValue = 'originalDataPointValue',
```
cant we have alphabetical order?

minotogna commented on src/client/store/pages/assessmentSection/slice.ts:
```diff
@@ -60,7 +59,11 @@ export const assessmentSectionSlice = createSlice({
     builder.addCase(getOriginalDataPointData.fulfilled, (state, { payload }) => {
-      state.originalDataPointData = payload
+      const countryIso = Object.keys(payload || {})[0] as CountryIso
+      if (countryIso) {
```
why shouldn't we have countryIso here?

sorja committed:
minor code improvements

minotogna commented on src/test/integration/assessment/originalDataPoint.ts:
```diff
@@ -47,7 +50,9 @@ export default (): void =>
     it('Edit existing/not existing Original data point', async () => {
       const editedOriginalDataPoint = await CycleDataController.updateOriginalDataPoint({
-        assessment,
+        assessment: {
+          ...assessment,
+        },
```
why just spreading itself ? is it needed ?

sorja replied:
typo

minotogna commented on src/server/repository/assessment/assessment/createAssessment.ts:
```diff
@@ -7,15 +7,17 @@ import { read } from '@server/repository/assessment/assessment/read'
 export const createAssessment = async (
   params: {
-    assessment: Pick<Assessment, 'props'>
+    assessment: Pick<Assessment, 'props' | 'metaCache'>
```
when creating an assessment, metaCache is not populated. 
can u avoid passing metaCache here?

sorja replied:
how should we populate the test assessment?
we fetch it from db and if we dont add it there, dont add it here, then.. what is left?

minotogna replied:
check my pr

mergify[bot] committed:
Merge branch 'master-meta' into meta/update-dependants-on-odp-update
